### ISSUE TITLE
Distributed Processing in any way?

### ISSUE THREAD

**Issue by LeaveNhA** [ml-explore/mlx-examples#374]
Hello, as you might know, I'm admiring your works (all of you guys, all the contributors) and love our community.

Apart from this start, here is my simple question:

Is there any plan to make it distributed or can I use any already written library/frameworks that I can use for this purpose?

Thank you,
Good luck!

**Comment by awni** [comment]
Could you say more about what you are looking for? Distributed is a pretty loaded term.

What exactly would you like to distribute? Training / inference? At what granularity?

**Comment by LeaveNhA** [comment]
Let's assume I have a couple of high-end apple device on my home. I wanna use them together to generate more processing power. 
Both for training and inference.

**Comment by danilopeixoto** [comment]
We could be referring to solutions like [Spark](https://spark.apache.org/docs/latest/cluster-overview.html) or [Dask](https://docs.dask.org/en/stable/deploying.html) (with local, Kubernetes, MPI, and other backends) for distributed data processing (that could eventually implement ML specifics, such as [Dask-ML](https://ml.dask.org)) or something very specific to ML from starting point like [PyTorch](https://pytorch.org/docs/stable/distributed.html) and [TensorFlow](https://www.tensorflow.org/guide/distributed_training) distributed training features.

# Implementation examples

## Dask

### Set up

The setup depends on backend. In this example, we are defining a local cluster.

Start manager (scheduler):

```console
dask scheduler
```

Start workers:

```console
dask worker <dask-scheduler-address>
```

The cluster can also be created and managed using Python:

```python
from dask.distributed import LocalCluster


# The cluster object allows to scale the number of workers remotely
cluster = LocalCluster()
```

The workers need to discover the manager in the network and share access to resources such as files and data sources.

### Usage

Standalone:

```python
import dask.dataframe as dd


df = dd.read_csv(...)
df.x.sum().compute()
```

Local cluster:

```python
from dask.distributed import Client
import dask.dataframe as dd


client = Client('<dask-scheduler-address>')

# Use the default client object from runtime
df = dd.read_csv(...)
df.x.sum().compute()
```

### ML usage

Local cluster:

```python
from dask.distributed import Client
import dask.dataframe as dd

from dask_ml.cluster import KMeans


client = Client('<dask-scheduler-address>')

# Use the default client object from runtime
df = dd.read_csv(...)

kmeans = KMeans(n_clusters=3, init_max_iter=2, oversampling_factor=10)
kmeans.fit(df.to_dask_array(lengths=True))
kmeans.predict(df.head()).compute()
```

**Comment by LeaveNhA** [comment]
Man, you really enlighten the path. 
Yes, we can. Give me a couple of days to read the documentation and implementation details. 

I have a question in advance, how far we have to dive in order to implement such a backend? Can we find best practises or can we get ideas of from other backends (I assume they have other backends). 

Thank you, trully,
Sincerely.

**Comment by LeaveNhA** [comment]
I think, while we are busy, this problem solved, right?

**Comment by danilopeixoto** [comment]
It seems that MLX team added MPI training support.

**Comment by awni** [comment]
We did indeed: https://ml-explore.github.io/mlx/build/html/usage/distributed.html

I think we can close this and open more targeted issues related to distributed models as they come up.
